provider/oracle/environ.go

```diff
@@ -0,0 +1,870 @@
+	// anything.
+	// gsamfira: I could not find a CentOS image in the oracle compute
+	// marketplace, so not doing anything CentOS specific here.
+	var scripts []string
+	switch operatingSystem {
+	case os.Ubuntu:
+		for key, _ := range networks {
+			if key == defaultNicName {
+				continue
+			}
+			fileBaseName := fmt.Sprintf("%s.cfg", key)
+			fileContents := fmt.Sprintf(ubuntuInterfaceTemplate, key, key)
+			fileName := renderer.Join(
+				renderer.FromSlash(interfacesConfigDir), fileBaseName)
+			scripts = append(scripts,
+				renderer.WriteFile(fileName, []byte(fileContents))...)
+			scripts = append(scripts, fmt.Sprintf("/sbin/ifup %s", key))
+			scripts = append(scripts, "")
+
+		}
+		if len(scripts) > 0 {
+			cloudcfg.AddScripts(strings.Join(scripts, "\n"))
+		}
+	}
+	return cloudcfg, nil
+}
+
+func (e *oracleEnviron) getInstanceNetworks(
+	args environs.StartInstanceParams,
+	secLists, vnicSets []string,
+) (map[string]oci.Networker, error) {
+
+	// gsamfira: We add a default NIC attached o the shared network provided
+	// by the oracle cloud. This NIC is used for outbound traffic.
+	// While you can attach just a VNIC to the instance, and assign a
+	// public IP to it, I have not been able to get any outbound traffic
+	// through the thing.
+	// NOTE (gsamfira): The NIC ordering inside the instance is determined
+	// by the order in which the API returns the network interfaces. It seems
+	// the API orders the NICs alphanumerically. When adding new network
+	// interfaces, make sure they are ordered to come after eth0
+	networking := map[string]oci.Networker{
+		defaultNicName: oci.SharedNetwork{
+			Seclists: secLists,
+			Name_servers: []string{
+				"8.8.8.8",
+				"8.8.4.4",
+			},
+		},
+	}
+	spaces := map[string]bool{}
+	if len(args.EndpointBindings) != 0 {
+		for _, spaceProviderID := range args.EndpointBindings {
+			logger.Infof("Adding space %s", string(spaceProviderID))
+			spaces[string(spaceProviderID)] = true
+		}
+	}
+	if s := args.Constraints.IncludeSpaces(); len(s) != 0 {
+		for _, val := range s {
+			logger.Infof("Adding space %s", val)
+			spaces[val] = true
+		}
+	}
+	// No spaces specified by user. Just return the default NIC
+	if len(spaces) == 0 {
+		return networking, nil
+	}
+
+	providerSpaces, err := e.getIPExchangesAndNetworks()
+	if err != nil {
+		return map[string]oci.Networker{}, err
+	}
+	//start from 1. eth0 is the default nic that gets attached by default.
+	idx := 1
+	logger.Infof("have spaces %v", spaces)
+	for space, _ := range spaces {
+		providerID := e.client.ComposeName(space)
+		providerSpace, ok := providerSpaces[providerID]
+		if !ok {
+			return map[string]oci.Networker{},
+				errors.Errorf("Could not find space %q", space)
+		}
+		if len(providerSpace) == 0 {
+			return map[string]oci.Networker{},
+				errors.Errorf("No usable subnets found in space %q", space)
+		}
+
+		// gsamfira: about as random as rainfall during monsoon season.
```

Comment:
@hoenirvili  no need to tag me more then once. I will address all comments as I get to that bit of the code :).